src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs

```diff
@@ -726,5 +727,91 @@ void M(delegate*<delegate*<object, void>, void> param1, delegate*<delegate*<obje
                 Diagnostic(ErrorCode.ERR_NoImplicitConvCast, "param2").WithArguments("delegate*<delegate*<object>>", "delegate*<delegate*<string>>").WithLocation(7, 45)
             );
         }
+
+        [Fact]
+        public void FunctionPointerParameterTypeInference()
+        {
+            var comp = CreateCompilationWithFunctionPointers(@"
+unsafe class C
+{
+    public void M1<T>(delegate*<T, void> param) {}
+    public void M2()
+    {
+        delegate*<string, void> p = null;
+        M1(p);
+    }
+}");
+
+            comp.VerifyDiagnostics();
+
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+
+            var invocation = tree.GetRoot().DescendantNodes().OfType<InvocationExpressionSyntax>().Single();
+            var methodSymbol = (IMethodSymbol)model.GetSymbolInfo(invocation).Symbol!;
+
+            Assert.NotSame(methodSymbol, methodSymbol.OriginalDefinition);
+            Assert.Equal(SpecialType.System_String, methodSymbol.TypeArguments[0].SpecialType);
+            var functionPointer = (FunctionPointerTypeSymbol)methodSymbol.Parameters[0].Type;
```

Comment:
>(FunctionPointerTypeSymbol) [](start = 34, length = 27)

This feels wrong. How could this cast succeed on an ITypeSymbol?

Reply:
This is a symbol from the `PublicModel` namespace. I can qualify it to make it obvious.

---
In reply to: [406417601](https://github.com/dotnet/roslyn/pull/43041#discussion_r406417601) [](ancestors = 406417601)